reps/2023-02-03-node-affinity-feature-enhancements.md

```diff
@@ -0,0 +1,437 @@
+  absl::flat_hash_map<string, absl::flat_hash_map<string, absl::flat_hash_map<string, absl::flat_hash_map<string, int>>>> labels;
+}
+```
+
+3. When a node registers with GCS and broadcasts to all nodes, add the static labels information to NodeResources.  
+
+Because NodeInfo originally had a mechanism to synchronize to the entire cluster. So there is no need to modify too much code.
+
+4. Scheduling optimization through Labels 
+
+Now any node raylet has node static labels information for all nodes.  
+when NodeAffinity schedules, if it traverses the Labels of each node, the algorithm complexity is very large, and the performance will be poor.   
+<b> Therefore, it is necessary to generate a full-cluster node labels index table to improve scheduling performance. </b>
+
+The label index table is only used to obtain which nodes have this key and value. The complexity of this algorithm is O(1).
+The matching logic of the expression is as follows:
+> label_in(key, values[]) = LabelManager.GetNodesByKeyAndValue()
+
+> label_not_in(key, values[]) = All_Nodes - LabelManager.GetNodesByKeyAndValue()
+
+```
+class ClusterLabelManager {
+ public:
+  absl::flat_hash_set<NodeID> GetNodesByNodeKeyAndValue(const std::string &key, const absl::flat_hash_set<std::string> &values) const;
+
+  absl::flat_hash_set<NodeID> GetNodesByNodeKey(const std::string &key) const;
+
+  void AddNodeLabels(const std::shared_ptr<NodeInfo> &node);
+
+  void RemoveNodeLabels(const std::shared_ptr<NodeInfo> &node);
+
+ private:
+ <namespace, <label_key, <lable_value, <node_id, ref_count>>>> labels_to_nodes_;
+}
+```
+
+### If the Match Expression Cannot be satisfied
+If the matching expression cannot be satisfied, The actor/task will be add to the pending queue. Util the matching expression all be statisfied.
+
+1. resources are enough, but node_id of affinity don't exist. -> throw exception, compatible with current behavior 
+
+2. resources are enough, but labels affinity  cannot be satisfied? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+3. labels affinity can be satisfied, but resources are not enough? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+4. both labelsaffinity and resources are not satisfied? -> Hang & report schedule failed event and detail unstaisfed reason to exposed to users
+
+
+### how other system achieve the same goal?
+1、K8s
+This solution is to learn the PodAffinity/NodeAffinity features of K8s。
+https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#affinity-and-anti-affinity
+
+Scheduling Policy | Label Owner | Operator
+-- | -- | --
+nodeAffinity | NODE | In, NotIn, Exists, DoesNotExist, Gt, Lt
+podAffinity | POD | In, NotIn, Exists, DoesNotExist
+podAntiAffinity | POD | In, NotIn, Exists, DoesNotExist
+
+
+### what's the alternative to achieve the same goal?
+** Option 3: putting Labels into the custom resource solution  **  
```

Comment:
Not necessarily saying that I'm proposing this but have we considered label being a k-v pair vs a string. Does a string label have the same expressiveness as a k-v label? A good thing of string label is that it can be easily implemented as custom resources internally.

Reply:
@jjyao 

>have we considered label being a k-v pair vs a string?
We should have discussed this before, and now the form of key-value should refer to the best form of other excellent experience.

1. The (key-value) labels mechanism of K8s is one of the highlights of K8s and has been recognized by countless people.
2. The key-value is simple and can be accepted by users naturally. But at the same time a certain flexibility can be submitted. For example: {gpu_type: "A100"} , {gpu_type: "T100"} . This setting is very natural and intuitive. But if it is replaced by a string: {"gpu_type@A100"},{"gpu_type@T100"}. It will be a little troublesome for users.

> implemented as custom resources internally

This issue has been discussed before. NodeAffinity uses custom_resource internally. Wouldn't be a good solution. It not only destroys the semantics of the original custom_resouce, but also the code implementation will be more complicated (1+1)> 2.

https://github.com/ray-project/enhancements/pull/13/files?short_path=5ae2129#diff-5ae21290dd6f2483719e2557543e9335c1867bea3d1f1cfe516e269d0bd9bc12

![image](https://user-images.githubusercontent.com/11072802/221743480-5b6c13fa-c4ba-4335-9495-86a53589675b.png)